File: agent/team/manager.py
from __future__ import annotations

import asyncio
import json
from collections.abc import Awaitable, Callable
from pathlib import Path
from threading import Lock
from typing import Any

from loguru import logger

from ..providers.base import run_sync
from ..subagents import SubagentSpec
from ..tasks import TaskKind
from ..tools.registry import ToolRegistry
from . import events
from .bus import MessageBus
from .models import (
    LEAD_ACTOR,
    TeamMember,
    TeamMessage,
    TeamStatus,
    new_id,
    now_ts,
    validate_member_name,
)
from .store import TeamStore

StreamEmitter = Callable[[dict[str, Any]], Awaitable[None]]


_ROLE_AGENT_TYPES = {
    "coder": "neiguan_yingzao",
    "reviewer": "shangbao_dianbu",
    "researcher": "dongchang_tanshi",
    "reader": "sili_suitang",
    "runner": "xiaohuangmen",
}


def role_to_agent_type(role: str) -> str:
    return _ROLE_AGENT_TYPES.get(str(role or "").strip().lower(), "sili_suitang")


class TeamManager:
    def __init__(
        self,
        *,
        root,
        team_dir: str | Path | None = None,
        project_id: str | None = None,
        parent_registry: ToolRegistry,
        subagent_registry,
        runner_factory,
        task_manager=None,
    ):
        self.project_id = str(project_id or "").strip() or None
        self.store = TeamStore(root, team_dir=Path(team_dir) if team_dir is not None else None)
        self.bus = MessageBus(self.store)
        self.parent_registry = parent_registry
        self.subagent_registry = subagent_registry
        self.runner_factory = runner_factory
        self.task_manager = task_manager
        self._locks: dict[str, Lock] = {}
        self._locks_guard = Lock()

    def payload(self) -> dict[str, Any]:
        config = self.store.load_config()
        members = []
        for member in self.store.list_members():
            item = member.to_dict()
            item["unread"] = self.bus.unread_count(member.name)
            item["recent_messages"] = [msg.to_dict() for msg in self.bus.recent(member.name, limit=5)]
            item["thread_count"] = len(self.store.read_thread(member.name))
            item["tools"] = self._tool_names_for_member(member)
            members.append(item)
        return {
            "config": config,
            "members": members,
            "leadUnread": self.bus.unread_count(LEAD_ACTOR),
            "leadInbox": [msg.to_dict() for msg in self.bus.recent(LEAD_ACTOR, limit=50)],
        }

    def member_payload(self, name: str) -> dict[str, Any]:
        member = self._require_member(name)
        return {
            "member": {
                **member.to_dict(),
                "unread": self.bus.unread_count(member.name),
                "tools": self._tool_names_for_member(member),
            },
            "inbox": [msg.to_dict() for msg in self.bus.recent(member.name, limit=100)],
            "leadInbox": [msg.to_dict() for msg in self.bus.recent(LEAD_ACTOR, limit=100)],
            "thread": self._thread_summary(member.name),
        }

    def spawn_teammate(
        self,
        *,
        name: str,
        role: str,
        task: str | None = None,
        agent_type: str | None = None,
        sender: str = LEAD_ACTOR,
        emit: StreamEmitter | None = None,
        loop: asyncio.AbstractEventLoop | None = None,
        parent_call_id: str | None = None,
    ) -> str:
        safe_name = validate_member_name(name)
        resolved_agent_type = agent_type or role_to_agent_type(role)
        spec = self.subagent_registry.get(resolved_agent_type)
        if spec is None:
            return (
                f"Error: unknown agent_type '{resolved_agent_type}'. "
                f"Available: {self.subagent_registry.names(include_aliases=True)}"
            )

        existing = self.store.get_member(safe_name)
        if existing and existing.status == TeamStatus.SHUTDOWN.value:
            existing = existing.touch(status=TeamStatus.IDLE.value, last_error=None)
        member = TeamMember(
            name=safe_name,
            role=role,
            agent_type=self.subagent_registry.resolve_name(resolved_agent_type),
            status=(existing.status if existing else TeamStatus.IDLE.value),
            created_at=(existing.created_at if existing else now_ts()),
            last_error=(existing.last_error if existing else None),
        )
        self.store.upsert_member(member)
        self._emit(events.member_update(member), emit, loop)

        if not task:
            return json.dumps({"created": member.to_dict()}, ensure_ascii=False)

        task_id = new_id("task")
        msg = self.bus.send(
            from_actor=sender,
            to=member.name,
            content=task,
            type="task",
            task_id=task_id,
        )
        self._emit(events.message_event(msg), emit, loop)
        result = self.wake_teammate(
            member.name,
            emit=emit,
            loop=loop,
            parent_call_id=parent_call_id,
            purpose=task[:120],
        )
        return json.dumps(
            {"created": member.to_dict(), "message": msg.to_dict(), "result": result},
            ensure_ascii=False,
        )

    def list_teammates(self) -> str:
        return json.dumps(self.payload(), ensure_ascii=False, indent=2)

    def read_inbox(self, *, actor: str = LEAD_ACTOR, limit: int = 20, mark_read: bool = True) -> str:
        messages = self.bus.read(actor, limit=limit, mark_read=mark_read)
        return json.dumps([message.to_dict() for message in messages], ensure_ascii=False, indent=2)

    def send_message(
        self,
        *,
        to: str,
        content: str,
        sender: str = LEAD_ACTOR,
        wake: bool = True,
        type: str = "message",
        emit: StreamEmitter | None = None,
        loop: asyncio.AbstractEventLoop | None = None,
        parent_call_id: str | None = None,
    ) -> str:
        if to != LEAD_ACTOR:
            self._require_member(to)
        if sender != LEAD_ACTOR:
            self._require_member(sender)
        msg = self.bus.send(from_actor=sender, to=to, content=content, type=type)
        self._emit(events.message_event(msg), emit, loop)
        result = None
        if wake and to != LEAD_ACTOR:
            result = self.wake_teammate(
                to,
                emit=emit,
                loop=loop,
                parent_call_id=parent_call_id,
                purpose=content[:120],
            )
        return json.dumps({"message": msg.to_dict(), "result": result}, ensure_ascii=False)

    def broadcast(
        self,
        *,
        content: str,
        recipients: list[str] | None = None,
        wake: bool = True,
        emit: StreamEmitter | None = None,
        loop: asyncio.AbstractEventLoop | None = None,
        parent_call_id: str | None = None,
    ) -> str:
        members = [
            member
            for member in self.store.list_members()
            if member.status != TeamStatus.SHUTDOWN.value
        ]
        if recipients:
            wanted = {validate_member_name(name) for name in recipients}
            members = [member for member in members if member.name in wanted]
        sent = []
        results = []
        for member in members:
            msg = self.bus.send(from_actor=LEAD_ACTOR, to=member.name, content=content, type="message")
            sent.append(msg.to_dict())
            self._emit(events.message_event(msg), emit, loop)
            if wake:
                results.append({
                    "name": member.name,
                    "result": self.wake_teammate(
                        member.name,
                        emit=emit,
                        loop=loop,
                        parent_call_id=parent_call_id,
                        purpose=content[:120],
                    ),
                })
        return json.dumps({"sent": sent, "results": results}, ensure_ascii=False, indent=2)

    def shutdown_teammate(self, *, name: str, emit: StreamEmitter | None = None,
                          loop: asyncio.AbstractEventLoop | None = None) -> str:
        member = self.store.update_member(
            name,
            status=TeamStatus.SHUTDOWN.value,
            last_error=None,
        )
        self._emit(events.member_update(member), emit, loop)
        return json.dumps({"shutdown": member.to_dict()}, ensure_ascii=False)

    def wake_teammate(
        self,
        name: str,
        *,
        emit: StreamEmitter | None = None,
        loop: asyncio.AbstractEventLoop | None = None,
        parent_call_id: str | None = None,
        purpose: str = "",
    ) -> str:
        member = self._require_member(name)
        if member.status == TeamStatus.SHUTDOWN.value:
            return f"Error: teammate '{member.name}' is shutdown"
        lock = self._lock_for(member.name)
        if not lock.acquire(blocking=False):
            return f"Error: teammate '{member.name}' is already working"
        try:
            return self._wake_locked(
                member,
                emit=emit,
                loop=loop,
                parent_call_id=parent_call_id,
                purpose=purpose,
            )
        finally:
            lock.release()

    def _wake_locked(
        self,
        member: TeamMember,
        *,
        emit: StreamEmitter | None,
        loop: asyncio.AbstractEventLoop | None,
        parent_call_id: str | None,
        purpose: str,
    ) -> str:
        working = self.store.update_member(
            member.name,
            status=TeamStatus.WORKING.value,
            last_error=None,
        )
        self._emit(events.member_update(working), emit, loop)
        self._emit(events.run_start(parent_id=parent_call_id, member=working, purpose=purpose), emit, loop)

        checkpoint = self.store.read_checkpoint_payload(working.name)
        pending_cursor_start: int | None = None
        pending_cursor_end: int | None = None
        pending_message_ids: list[str] = []
        if checkpoint:
            history = checkpoint["messages"]
            pending_cursor_start = checkpoint.get("pending_cursor_start")
            pending_cursor_end = checkpoint.get("pending_cursor_end")
            pending_message_ids = list(checkpoint.get("pending_message_ids") or [])
            inbox_by_id = {msg.id: msg for msg in self.bus.all_messages(working.name)}
            unread = [inbox_by_id[msg_id] for msg_id in pending_message_ids if msg_id in inbox_by_id]
        else:
            inbox = self.bus.all_messages(working.name)
            pending_cursor_start = min(self.store.read_cursor(working.name), len(inbox))
            unread = inbox[pending_cursor_start:pending_cursor_start + 50]
            pending_cursor_end = pending_cursor_start + len(unread)
            pending_message_ids = [msg.id for msg in unread]
            history = self.store.read_thread(working.name)

        if not checkpoint and not unread:
            idle = self.store.update_member(working.name, status=TeamStatus.IDLE.value, last_error=None)
            self._emit(events.member_update(idle), emit, loop)
            self._emit(events.run_done(parent_id=parent_call_id, member=idle, summary="没有未读消息。"), emit, loop)
            return "没有未读消息。"

        if not checkpoint:
            history.append({"role": "user", "content": self._render_inbox_for_runner(working, unread)})
        self.store.write_checkpoint(
            working.name,
            history,
            pending_cursor_start=pending_cursor_start,
            pending_cursor_end=pending_cursor_end,
            pending_message_ids=pending_message_ids,
        )

        task_record = None
        if self.task_manager is not None:
            task_record = self.task_manager.start_task(
                kind=TaskKind.TEAM_WAKE.value,
                title=f"Team wake: {working.name}",
                source="team",
                metadata={
                    "project_id": self.project_id or "default",
                    "member": working.name,
                    "role": working.role,
                    "agent_type": working.agent_type,
                    "parent_call_id": parent_call_id,
                    "pending_message_ids": list(pending_message_ids),
                    "resumed_from_checkpoint": bool(checkpoint),
                },
            )
            user_content = self._latest_user_content(history)
            if user_content:
                self.task_manager.append_sidechain(task_record.id, {
                    "role": "user",
                    "content": user_content,
                    "metadata": {
                        "member": working.name,
                        "pending_message_ids": list(pending_message_ids),
                    },
                })

        spec = self._require_spec(working.agent_type)
        sub_registry = self._registry_for_member(working, spec)
        runner = self.runner_factory(member=working, spec=spec, sub_registry=sub_registry)
        lead_before_ids = {msg.id for msg in self.bus.all_messages(LEAD_ACTOR)}

        async def team_emit(evt: dict[str, Any]) -> None:
            evt_type = str(evt.get("event") or "")
            if evt_type.startswith("team_"):
                if emit is None:
                    return
                if loop is not None:
                    self._emit(evt, emit, loop)
                    return
                await emit(evt)
                return
            mapped = self._map_runner_event(evt, working, parent_call_id)
            if mapped:
                if emit is None:
                    return
                if loop is not None:
                    self._emit(mapped, emit, loop)
                    return
                await emit(mapped)

        try:
            if emit is not None:
                final = run_sync(runner.step_stream(history, team_emit))
            else:
                final = runner.step(history)
            self.store.write_thread(working.name, history)
            self.store.clear_checkpoint(working.name)
            if pending_cursor_end is not None:
                self.store.write_cursor(working.name, pending_cursor_end)
            idle = self.store.update_member(working.name, status=TeamStatus.IDLE.value, last_error=None)
            self._emit(events.member_update(idle), emit, loop)
            explicit_reply = any(
                msg.id not in lead_before_ids and msg.from_actor == working.name
                for msg in self.bus.all_messages(LEAD_ACTOR)
            )
            if not explicit_reply:
                result_msg = self.bus.send(
                    from_actor=working.name,
                    to=LEAD_ACTOR,
                    content=final,
                    type="result",
                    in_reply_to=(pending_message_ids[-1] if pending_message_ids else None),
                    meta={"role": working.role, "agent_type": working.agent_type},
                )
                self._emit(events.message_event(result_msg), emit, loop)
            if task_record is not None and self.task_manager is not None:
                self.task_manager.append_sidechain(task_record.id, {
                    "role": "assistant",
                    "content": final,
                    "metadata": {"member": working.name},
                })
                self.task_manager.complete_task(task_record.id, summary=final[:500])
            logger.info(f"[队友回禀 · {working.name}]: {final[:500]}")
            return final
        except Exception as exc:
            err = str(exc)
            logger.exception(f"team wake failed: {working.name}")
            self.store.write_checkpoint(
                working.name,
                history,
                pending_cursor_start=pending_cursor_start,
                pending_cursor_end=pending_cursor_end,
                pending_message_ids=pending_message_ids,
            )
            error_member = self.store.update_member(
                working.name,
                status=TeamStatus.ERROR.value,
                last_error=err,
            )
            self._emit(events.member_update(error_member), emit, loop)
            self._emit(events.run_error(parent_id=parent_call_id, member=error_member, message=err), emit, loop)
            error_msg = self.bus.send(
                from_actor=working.name,
                to=LEAD_ACTOR,
                content=err,
                type="error",
                meta={"role": working.role, "agent_type": working.agent_type},
            )
            self._emit(events.message_event(error_msg), emit, loop)
            if task_record is not None and self.task_manager is not None:
                self.task_manager.append_sidechain(task_record.id, {
                    "role": "error",
                    "content": err,
                    "metadata": {"member": working.name},
                })
                self.task_manager.fail_task(task_record.id, error=err)
            return f"Error: teammate '{working.name}' raised: {err}"

    def _registry_for_member(self, member: TeamMember, spec: SubagentSpec) -> ToolRegistry:
        from .tools import TeamReadInboxTool, TeamSendMessageTool

        registry = ToolRegistry()
        for tool_name in spec.tool_names:
            tool = self.parent_registry.get(tool_name)
            if tool is not None:
                registry.register(tool)
        registry.register(TeamSendMessageTool(self, sender=member.name, allow_wake=False))
        registry.register(TeamReadInboxTool(self, actor=member.name))
        return registry

    def _tool_names_for_member(self, member: TeamMember) -> list[str]:
        spec = self.subagent_registry.get(member.agent_type)
        if spec is None:
            return []
        return [*spec.tool_names, "send_message", "read_inbox"]

    def _thread_summary(self, name: str) -> list[dict[str, Any]]:
        out = []
        for item in self.store.read_thread(name)[-20:]:
            content = item.get("content", "")
            if isinstance(content, list):
                content = "".join(
                    str(block.get("text", ""))
                    for block in content
                    if isinstance(block, dict) and block.get("type") == "text"
                )
            out.append({
                "role": item.get("role"),
                "content": str(content)[:2000],
            })
        return out

    @staticmethod
    def _latest_user_content(history: list[dict[str, Any]]) -> str:
        for item in reversed(history):
            if item.get("role") == "user":
                return str(item.get("content") or "")
        return ""

    def _require_member(self, name: str) -> TeamMember:
        member = self.store.get_member(name)
        if member is None:
            raise ValueError(f"unknown teammate: {name}")
        return member

    def _require_spec(self, agent_type: str) -> SubagentSpec:
        spec = self.subagent_registry.get(agent_type)
        if spec is None:
            raise ValueError(f"unknown agent_type: {agent_type}")
        return spec

    def _lock_for(self, name: str) -> Lock:
        safe = validate_member_name(name)
        with self._locks_guard:
            if safe not in self._locks:
                self._locks[safe] = Lock()
            return self._locks[safe]

    @staticmethod
    def _render_inbox_for_runner(member: TeamMember, messages: list[TeamMessage]) -> str:
        lines = [
            f"你是 Agent Team 队友 {member.name}，role={member.role}，agent_type={member.agent_type}。",
            "下面是你的未读 inbox。请处理这些消息，必要时调用工具，最后用 send_message(to=\"lead\", content=\"...\") 回禀，随后给出简短总结。",
            "",
            "## Inbox",
        ]
        for msg in messages:
            lines.append(
                f"- id={msg.id} type={msg.type} from={msg.from_actor} "
                f"task_id={msg.task_id or ''}: {msg.content}"
            )
        return "\n".join(lines)

    @staticmethod
    def _map_runner_event(evt: dict[str, Any], member: TeamMember,
                          parent_call_id: str | None) -> dict[str, Any] | None:
        evt_type = evt.get("event")
        if evt_type == "message_delta":
            return events.run_delta(parent_id=parent_call_id, member=member, delta=str(evt.get("delta") or ""))
        if evt_type == "tool_call":
            return events.run_tool_call(
                parent_id=parent_call_id,
                member=member,
                id=evt.get("id"),
                name=str(evt.get("name") or ""),
                arguments=evt.get("arguments") if isinstance(evt.get("arguments"), dict) else {},
            )
        if evt_type == "tool_result":
            return events.run_tool_result(
                parent_id=parent_call_id,
                member=member,
                id=evt.get("id"),
                name=evt.get("name"),
                summary=str(evt.get("summary") or ""),
            )
        if evt_type == "tool_error":
            return events.run_tool_error(
                parent_id=parent_call_id,
                member=member,
                id=evt.get("id"),
                name=evt.get("name"),
                message=str(evt.get("message") or ""),
            )
        if evt_type == "assistant_done":
            return events.run_done(parent_id=parent_call_id, member=member, summary=str(evt.get("content") or ""))
        return None

    def _emit(
        self,
        event: dict[str, Any],
        emit: StreamEmitter | None,
        loop: asyncio.AbstractEventLoop | None,
    ) -> None:
        if emit is None:
            return
        if self.project_id and str(event.get("event") or "").startswith("team_"):
            event = {**event, "project_id": self.project_id}
        if loop is not None:
            asyncio.run_coroutine_threadsafe(emit(event), loop)
            return
        try:
            run_sync(emit(event))
        except RuntimeError:
            logger.debug(f"team event dropped outside loop: {event.get('event')}")

File: packages/core/src/team/manager.ts
import { ToolRegistry } from '../tools/registry'
import { MessageBus } from './bus'
import * as events from './events'
import { LEAD_ACTOR, TeamMember, TeamMessage, TeamStatus, newTeamId, validateMemberName } from './models'
import { TeamStore } from './store'
import { TeamReadInboxTool, TeamSendMessageTool } from './tools'

const ROLE_AGENT_TYPES: Record<string, string> = {
  coder: 'neiguan_yingzao',
  reviewer: 'shangbao_dianbu',
  researcher: 'dongchang_tanshi',
  reader: 'sili_suitang',
  runner: 'xiaohuangmen',
}

export function roleToAgentType(role: string): string {
  return ROLE_AGENT_TYPES[String(role || '').trim().toLowerCase()] ?? 'sili_suitang'
}

export interface TeamSubagentSpec { name?: string; tool_names?: string[]; toolNames?: string[] }
export interface TeamSubagentRegistry {
  get(name: string): TeamSubagentSpec | null | undefined
  resolveName?(name: string): string
  names?(includeAliases?: boolean): string[]
}
export interface TeamRunner {
  step(history: Array<Record<string, unknown>>): string | Promise<string>
  stepStream?(history: Array<Record<string, unknown>>, emit: (event: Record<string, unknown>) => Promise<void>): Promise<string>
}
export type TeamRunnerFactory = (opts: { member: TeamMember; spec: TeamSubagentSpec; subRegistry: ToolRegistry }) => TeamRunner
export type TeamEventSink = (event: Record<string, unknown>) => Promise<void> | void

export class TeamManager {
  readonly projectId: string | null
  readonly store: TeamStore
  readonly bus: MessageBus
  readonly parentRegistry: ToolRegistry
  readonly subagentRegistry: TeamSubagentRegistry
  readonly runnerFactory: TeamRunnerFactory | null
  readonly eventSink: TeamEventSink | null
  private working = new Set<string>()

  constructor(opts: {
    root: string
    teamDir?: string | null
    projectId?: string | null
    parentRegistry?: ToolRegistry | null
    subagentRegistry: TeamSubagentRegistry
    runnerFactory?: TeamRunnerFactory | null
    eventSink?: TeamEventSink | null
  }) {
    this.projectId = opts.projectId?.trim() || null
    this.store = new TeamStore(opts.root, { teamDir: opts.teamDir ?? null })
    this.bus = new MessageBus(this.store)
    this.parentRegistry = opts.parentRegistry ?? new ToolRegistry()
    this.subagentRegistry = opts.subagentRegistry
    this.runnerFactory = opts.runnerFactory ?? null
    this.eventSink = opts.eventSink ?? null
  }

  payload(): Record<string, unknown> {
    const members = this.store.listMembers().map((member) => ({
      ...member.toDict(),
      unread: this.bus.unreadCount(member.name),
      recent_messages: this.bus.recent(member.name, { limit: 5 }).map((msg) => msg.toDict()),
      thread_count: this.store.readThread(member.name).length,
      tools: this.toolNamesForMember(member),
    }))
    return {
      config: this.store.loadConfig(),
      members,
      leadUnread: this.bus.unreadCount(LEAD_ACTOR),
      leadInbox: this.bus.recent(LEAD_ACTOR, { limit: 50 }).map((msg) => msg.toDict()),
    }
  }

  async spawnTeammate(opts: { name: string; role: string; task?: string | null; agent_type?: string | null; sender?: string; parent_call_id?: string | null; eventSink?: TeamEventSink | null }): Promise<string> {
    const safeName = validateMemberName(opts.name)
    const resolved = opts.agent_type || roleToAgentType(opts.role)
    const spec = this.subagentRegistry.get(resolved)
    if (!spec) return `Error: unknown agent_type '${resolved}'. Available: ${this.subagentRegistry.names?.(true) ?? []}`
    const existing = this.store.getMember(safeName)
    const agentType = this.subagentRegistry.resolveName?.(resolved) ?? spec.name ?? resolved
    const member = new TeamMember({
      name: safeName,
      role: opts.role,
      agent_type: agentType,
      status: existing && existing.status !== TeamStatus.SHUTDOWN ? existing.status : TeamStatus.IDLE,
      created_at: existing?.created_at,
      last_error: existing?.last_error ?? null,
    })
    this.store.upsertMember(member)
    await this.emit(events.memberUpdate(member), opts.eventSink)
    if (!opts.task) return JSON.stringify({ created: member.toDict() })

    const taskId = newTeamId('task')
    const msg = this.bus.send({ from_actor: opts.sender ?? LEAD_ACTOR, to: member.name, content: opts.task, type: 'task', task_id: taskId })
    await this.emit(events.messageEvent(msg), opts.eventSink)
    const result = await this.wakeTeammate(member.name, { parent_call_id: opts.parent_call_id ?? null, purpose: opts.task.slice(0, 120), eventSink: opts.eventSink ?? null })
    return JSON.stringify({ created: member.toDict(), message: msg.toDict(), result })
  }

  listTeammates(): string {
    return JSON.stringify(this.payload(), null, 2)
  }

  readInbox(opts: { actor?: string; limit?: number; mark_read?: boolean } = {}): string {
    const messages = this.bus.read(opts.actor ?? LEAD_ACTOR, { limit: opts.limit ?? 20, mark_read: opts.mark_read ?? true })
    return JSON.stringify(messages.map((msg) => msg.toDict()), null, 2)
  }

  async sendMessage(opts: { to: string; content: string; sender?: string; wake?: boolean; type?: string; parent_call_id?: string | null; eventSink?: TeamEventSink | null }): Promise<string> {
    if (opts.to !== LEAD_ACTOR) this.requireMember(opts.to)
    if ((opts.sender ?? LEAD_ACTOR) !== LEAD_ACTOR) this.requireMember(opts.sender ?? LEAD_ACTOR)
    const msg = this.bus.send({ from_actor: opts.sender ?? LEAD_ACTOR, to: opts.to, content: opts.content, type: opts.type ?? 'message' })
    await this.emit(events.messageEvent(msg), opts.eventSink)
    let result: string | null = null
    if ((opts.wake ?? true) && opts.to !== LEAD_ACTOR) result = await this.wakeTeammate(opts.to, { parent_call_id: opts.parent_call_id ?? null, purpose: opts.content.slice(0, 120), eventSink: opts.eventSink ?? null })
    return JSON.stringify({ message: msg.toDict(), result })
  }

  async broadcast(opts: { content: string; recipients?: string[] | null; wake?: boolean; parent_call_id?: string | null; eventSink?: TeamEventSink | null }): Promise<string> {
    let members = this.store.listMembers().filter((member) => member.status !== TeamStatus.SHUTDOWN)
    if (opts.recipients?.length) {
      const wanted = new Set(opts.recipients.map(validateMemberName))
      members = members.filter((member) => wanted.has(member.name))
    }
    const sent: Array<Record<string, unknown>> = []
    const results: Array<Record<string, unknown>> = []
    for (const member of members) {
      const msg = this.bus.send({ from_actor: LEAD_ACTOR, to: member.name, content: opts.content, type: 'message' })
      sent.push(msg.toDict())
      await this.emit(events.messageEvent(msg), opts.eventSink)
      if (opts.wake ?? true) results.push({ name: member.name, result: await this.wakeTeammate(member.name, { parent_call_id: opts.parent_call_id ?? null, purpose: opts.content.slice(0, 120), eventSink: opts.eventSink ?? null }) })
    }
    return JSON.stringify({ sent, results }, null, 2)
  }

  async shutdownTeammate(opts: { name: string; eventSink?: TeamEventSink | null }): Promise<string> {
    const member = this.store.updateMember(opts.name, { status: TeamStatus.SHUTDOWN, last_error: null })
    await this.emit(events.memberUpdate(member), opts.eventSink)
    return JSON.stringify({ shutdown: member.toDict() })
  }

  async wakeTeammate(name: string, opts: { parent_call_id?: string | null; purpose?: string; eventSink?: TeamEventSink | null } = {}): Promise<string> {
    const member = this.requireMember(name)
    if (member.status === TeamStatus.SHUTDOWN) return `Error: teammate '${member.name}' is shutdown`
    if (this.working.has(member.name)) return `Error: teammate '${member.name}' is already working`
    this.working.add(member.name)
    try {
      return await this.wakeLocked(member, opts)
    } finally {
      this.working.delete(member.name)
    }
  }

  private async wakeLocked(member: TeamMember, opts: { parent_call_id?: string | null; purpose?: string; eventSink?: TeamEventSink | null }): Promise<string> {
    const working = this.store.updateMember(member.name, { status: TeamStatus.WORKING, last_error: null })
    await this.emit(events.memberUpdate(working), opts.eventSink)
    await this.emit(events.runStart({ parent_id: opts.parent_call_id ?? null, member: working, purpose: opts.purpose ?? '' }), opts.eventSink)

    const inbox = this.bus.allMessages(working.name)
    const cursorStart = Math.min(this.store.readCursor(working.name), inbox.length)
    const unread = inbox.slice(cursorStart, cursorStart + 50)
    const cursorEnd = cursorStart + unread.length
    const pendingIds = unread.map((msg) => msg.id)
    const history = this.store.readThread(working.name)
    if (!unread.length) {
      const idle = this.store.updateMember(working.name, { status: TeamStatus.IDLE, last_error: null })
      await this.emit(events.memberUpdate(idle), opts.eventSink)
      await this.emit(events.runDone({ parent_id: opts.parent_call_id ?? null, member: idle, summary: '没有未读消息。' }), opts.eventSink)
      return '没有未读消息。'
    }

    history.push({ role: 'user', content: TeamManager.renderInboxForRunner(working, unread) })
    this.store.writeCheckpoint(working.name, history, { pending_cursor_start: cursorStart, pending_cursor_end: cursorEnd, pending_message_ids: pendingIds })
    if (!this.runnerFactory) throw new Error('team runner factory is unavailable')
    const spec = this.requireSpec(working.agent_type)
    const runner = this.runnerFactory({ member: working, spec, subRegistry: this.registryForMember(working, spec) })
    const leadBefore = new Set(this.bus.allMessages(LEAD_ACTOR).map((msg) => msg.id))

    try {
      const final = runner.stepStream ? await runner.stepStream(history, async (evt) => { await this.emit(this.mapRunnerEvent(evt, working, opts.parent_call_id ?? null) ?? evt, opts.eventSink) }) : await runner.step(history)
      this.store.writeThread(working.name, history)
      this.store.clearCheckpoint(working.name)
      this.store.writeCursor(working.name, cursorEnd)
      const idle = this.store.updateMember(working.name, { status: TeamStatus.IDLE, last_error: null })
      await this.emit(events.memberUpdate(idle), opts.eventSink)
      const explicitReply = this.bus.allMessages(LEAD_ACTOR).some((msg) => !leadBefore.has(msg.id) && msg.from_actor === working.name)
      if (!explicitReply) {
        const reply = this.bus.send({ from_actor: working.name, to: LEAD_ACTOR, content: final, type: 'result', in_reply_to: pendingIds.at(-1) ?? null, meta: { role: working.role, agent_type: working.agent_type } })
        await this.emit(events.messageEvent(reply), opts.eventSink)
      }
      await this.emit(events.runDone({ parent_id: opts.parent_call_id ?? null, member: idle, summary: final }), opts.eventSink)
      return final
    } catch (error) {
      const text = error instanceof Error ? error.message : String(error)
      this.store.writeCheckpoint(working.name, history, { pending_cursor_start: cursorStart, pending_cursor_end: cursorEnd, pending_message_ids: pendingIds })
      const errored = this.store.updateMember(working.name, { status: TeamStatus.ERROR, last_error: text })
      await this.emit(events.memberUpdate(errored), opts.eventSink)
      await this.emit(events.runError({ parent_id: opts.parent_call_id ?? null, member: errored, message: text }), opts.eventSink)
      return `Error: teammate '${working.name}' raised: ${text}`
    }
  }

  private registryForMember(member: TeamMember, spec: TeamSubagentSpec): ToolRegistry {
    const registry = new ToolRegistry()
    for (const name of toolNames(spec)) {
      const tool = this.parentRegistry.get(name)
      if (tool) registry.register(tool)
    }
    registry.register(new TeamSendMessageTool(this, { sender: member.name, allowWake: false }))
    registry.register(new TeamReadInboxTool(this, { actor: member.name }))
    return registry
  }

  private toolNamesForMember(member: TeamMember): string[] {
    const spec = this.subagentRegistry.get(member.agent_type)
    return spec ? [...toolNames(spec), 'send_message', 'read_inbox'] : []
  }

  private requireMember(name: string): TeamMember {
    const member = this.store.getMember(name)
    if (!member) throw new Error(`unknown teammate: ${name}`)
    return member
  }

  private requireSpec(agentType: string): TeamSubagentSpec {
    const spec = this.subagentRegistry.get(agentType)
    if (!spec) throw new Error(`unknown agent_type: ${agentType}`)
    return spec
  }

  private mapRunnerEvent(evt: Record<string, unknown>, member: TeamMember, parentId: string | null): Record<string, unknown> | null {
    const type = evt.event
    if (type === 'message_delta') return events.runDelta({ parent_id: parentId, member, delta: String(evt.delta ?? '') })
    if (type === 'tool_call') return events.runToolCall({ parent_id: parentId, member, id: stringOrNull(evt.id), name: String(evt.name ?? ''), arguments: isRecord(evt.arguments) ? evt.arguments : {} })
    if (type === 'tool_result') return events.runToolResult({ parent_id: parentId, member, id: stringOrNull(evt.id), name: stringOrNull(evt.name), summary: String(evt.summary ?? '') })
    if (type === 'tool_error') return events.runToolError({ parent_id: parentId, member, id: stringOrNull(evt.id), name: stringOrNull(evt.name), message: String(evt.message ?? '') })
    if (type === 'assistant_done') return events.runDone({ parent_id: parentId, member, summary: String(evt.content ?? '') })
    return null
  }

  static renderInboxForRunner(member: TeamMember, messages: TeamMessage[]): string {
    const lines = [
      `你是 Agent Team 队友 ${member.name}，role=${member.role}，agent_type=${member.agent_type}。`,
      '下面是你的未读 inbox。请处理这些消息，必要时调用工具，最后用 send_message(to="lead", content="...") 回禀，随后给出简短总结。',
      '',
      '## Inbox',
    ]
    for (const msg of messages) lines.push(`- id=${msg.id} type=${msg.type} from=${msg.from_actor} task_id=${msg.task_id ?? ''}: ${msg.content}`)
    return lines.join('\n')
  }

  private async emit(event: Record<string, unknown>, eventSink?: TeamEventSink | null): Promise<void> {
    const sink = eventSink ?? this.eventSink
    if (!sink) return
    const payload = this.projectId && String(event.event ?? '').startsWith('team_') ? { ...event, project_id: this.projectId } : event
    await sink(payload)
  }
}

function toolNames(spec: TeamSubagentSpec): string[] {
  return spec.tool_names ?? spec.toolNames ?? []
}

function stringOrNull(value: unknown): string | null {
  return value === undefined || value === null ? null : String(value)
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value))
}
